core/src/main/java/net/opentsdb/query/processor/bucketquantile/BucketQuantileNumericArrayProcessor.java

```diff
@@ -0,0 +1,348 @@
+          results[i] = new_value;
+        }
+      }
+      
+      if (limit < 0) {
+        LOG.warn("No time series or data were found at index " + index);
+        limit = 0;
+        return;
+      }
+      
+      // setup the ptiles
+      for (int i = 0; i < quantiles.length; i++) {
+        // TODO - pools
+        if (node.doubleArrayPool() != null) {
+          pooled_objects[i] = node.doubleArrayPool().claim(limit);
+        }
+        quantiles[i] = pooled_objects[i] == null ? new double[limit] :
+          (double[]) pooled_objects[i].object();
+      }
+      
+      // now that we have all of the counts in order we can start the iteration.
+      for (int i = 0; i < limit; i++) {
+        // load the accumulator
+        // TODO - overflow
+        long sum = -1;
+        int negatives = 0;
+        for (int x = 0; x < sources.length; x++) {
+          if (indices[x] < 0) {
+            // null
+            accumulator[x] = -1;
+            negatives++;
+          } else {
+            accumulator[x] = results[x][indices[x]++];
+            if (accumulator[x] < 0) {
+              negatives++;
+            } else if (sum < 0 || cumulative) {
+              sum = accumulator[x];
+            } else {
+              sum += accumulator[x];
+            }
+          }
+        }
+        
+        // check for too much bad data.
+        if (negatives > 0 && negative_threshold > 0) {
+          if (((double) negatives / (double) quantiles.length) * 100 >= 
+              negative_threshold) {
+            for (int y = 0; y < quantiles.length; y++) {
+              quantiles[y][i] = Double.NaN;
+            }
+            // skip processing
+            continue;
+          }
+        } else if (sum < 0) {
+          for (int y = 0; y < quantiles.length; y++) {
+            quantiles[y][i] = Double.NaN;
+          }
+          continue;
+        }
+        
+        // quantiles are sorted so we keep counting until we exceed the quantiles, 
+        // then advance.
+        for (int y = 0; y < qtiles.size(); y++) {
+          p_thresholds[y] = ((double) sum) * qtiles.get(y);
+        }
+        
+        int threshold_idx = 0;
+        double so_far = -1;
+        int last_non_zero_bucket = -1;
+        OuterLoop:
+        for (int z = 0; z < sources.length; z++) {
+          if (accumulator[z] <= 0) {
+            // nan or null.
+            continue;
+          }
+          
+          if (cumulative) {
+            so_far = accumulator[z];
+          } else {
+            so_far += accumulator[z];
+          }
+          last_non_zero_bucket = z;
+          while (so_far >= p_thresholds[threshold_idx]) {
+            quantiles[threshold_idx][i] = 
+                ((BucketQuantile) node).buckets()[z].report;
+            p_thresholds[threshold_idx] = -Double.MIN_VALUE;
+            threshold_idx++;
+            if (threshold_idx >= quantiles.length) {
+              // done, no more work to do.
+              break OuterLoop;
```

Comment:
...yup.

Reply:
(For the record, I'm okay with jumps as long as they are clearer than the alternative.)